docs/rfc/0662-collaborative-objects.adoc

```diff
@@ -0,0 +1,690 @@
+* `history_type`, always `"automerge"`, this is here to allow for different
+  CRDT implementations in future.
+
+
+Each object is also created with a JSON schema. The schema is represented by an
+initial `schema.json` and a series of schema migrations which extend that
+initial schema. Schema migrations will not be addressed in detail in this RFC
+but we will show their feasibility.
+
+==== `change`
+
+This is the automerge change which this commit introduces. It is a binary file
+which must contain a single change and it's dependents must be the dependents
+referenced by the parents of the commit.
+
+
+=== Object IDs
+
+We require that there only be one root of the change graph we're replicating.
+Whilst we could merge change graphs with multiple roots this would be insecure.
+Imagine that an honest peer creates an object, if an attacker could create 
+another root node in the change graph and arrange for it to come before the
+honest root in a topological sort then the attacker could override the schema
+and other properties of the object.
+
+To have a single root we need an object ID which is derived from the attributes
+of the object, that way an attacker cannot manipulate the attributes of an
+object with the same ID. We achieve this by using the hash of the initial
+commit of the object as the ID.
+
+It is entirely plausible that a peer would create two distinct objects with
+identical initial states. Under many content addressing schemes this would lead
+to the two objects having the same ID, however, git Commits include a timestamp
+so this will not be a problem.
+
+
+=== Reconstructing Collaborative Objects
+
+Assuming we have replicated a number of collaborative objects from our tracking
+graph, we can now view the merged state of those objects. To do this we search
+through every `/cob/<typename>/<object ID>` reference for
+every remote we have and collect the change files for each object ID.
+
+At this point we have the hash linked graph of automerge changes, but we need
+to make sure that the merged document is authenticated and valid with respect
+to it's schema. To do this we start at the root of the hash graph and walk
+down the tree. As we encounter each change we check it's signature, apply it
+and check that the new document does not violate the schema. If it does violate
+the schema we discard the change and all dependent changes. Finally, we have an
+authenticated document which respects a given schema.
+
+It is important to note that this merging is at this point not stored in the
+repository - it can be performed in memory and may be cached. The result is
+that the user sees a single merged view of the object based on the contents
+of the remotes they have replicated. That is, there is no additional
+merge-then-commit step.
+
+=== Fetching Collaborative Objects
+
+Each time a repository creates a collaborative object tree it creates a ref
+pointing to that object at `refs/namespaces/<project>/cob/<typename>/<object ID>`, 
```

Comment:
There is one (scalability) aspect that is still not entirely clear to me: are collaborators on a specific object supposed to merge all changes by the other collaborators into their copy of the object? If so, when does this happen?

Just to clarify:

Suppose peers `P_a`, `P_b`, `P_c` are maintainers as per this document (for simplicity, let's leave out key delegations). `P_a` and `P_b` apply some changes to collaborative object `O` within the context of project `R`, while `P_c` does not. Peers `P_d` and `P_e` are tracking `R`.

Now, `P_e` establishes a network connection to `P_d` with the goal of retrieving any updates to `R`. Because of the rule that delegate peers (aka maintainers) must be replicated, the refs layout `P_d` presents looks like this:

```
refs/namespaces/R/refs/cob/frob/O
refs/namespaces/R/refs/remotes/P_a/cob/frob/O
refs/namespaces/R/refs/remotes/P_b/cob/frob/O
refs/namespaces/R/refs/remotes/P_c/cob/frob/O
```

Because commits can be applied in any order by the involved peers, those `O`s won't converge to the same commit SHA. The question I have is, do they converge to the same tree SHA? That is, do `P_{a..e}` all apply all changes they see to their own version of `O`? Or only if they themselves apply a change, which requires to record the dependencies?

In the latter case, I think `P_d`'s view would not exist (because only maintainers can collaborate under this proposal). But, `P_e` needs to collect past and present maintainers from `R`, and fetch all corresponding remote tracking branches.

Reply:
The intention here is that all known views of the CRDT are merged at query time. In the scenario you're laying out above `P_e` would merge the changes from `P_{a..c}` when querying and see that as the state of the object. This merge is not a git merge, there is no commit that corresponds to it, it's just the merge of all the views of the collaborative object that `P_e` knows about. If `P_e` contributes a change then this is the point at which they would create a "change commit", which records the dependencies of that change in the git commit graph.

"merge at query time" is a simplification, what we will most likely want to do is merge when receiving new changes and cache the merged state somewhere. The point is that it's not part of the git commit graph. Conceptually it's the same and allows for `P_{d,e}` to have a view on the object without having to record any changes.

Reply:
That makes sense, thanks.

So, when `ls-ref`ing `P_d`, `P_e` would see:

```
refs/namespaces/R/refs/remotes/P_a/cob/frob/O
refs/namespaces/R/refs/remotes/P_b/cob/frob/O
```

Which means we can't limit the response by `ref-prefix` :(

Reply:
Hmm, that's a good point. I guess we could do it if we had a separate replication step for collaborative objects once we've fetched the tracking graph?

Reply:
> once we've fetched the tracking graph

Right, that's the case anyway -- we would fetch `refs/namespaces/R/refs/remotes/P_a/cob/frob/O` if and only if it is mentioned in `refs/namespaces/R/refs/remotes/P_a/rad/signed_refs`, and only `want` the oid ibid.

What is more annoying is how to enable collaborators outside the inner cabal in the future: say you want to interrogate a seed node to see if it got changes by random peers for some object or class thereof. Or you would want to fetch the `cob/frob/O` ref of any random peer, without tracking that peer (ie. pruning it later if it turns out sketchy).

I suppose the only way to do this in a scalable way is to maintain some sort of reverse index by peer id, ie. `refs/namespaces/R/cob/frob/O/{self, P_a, P_b}`, where `P_*` are symrefs to the respective remote tracking heads.